bridge/jira/import.go

```diff
@@ -0,0 +1,613 @@
+				}
+				if err != nil {
+					out <- core.NewImportError(err, "")
+				}
+
+			}
+			if changelogIter.HasError() {
+				out <- core.NewImportError(changelogIter.Err, "")
+			}
+
+			if err := bug.CommitAsNeeded(); err != nil {
+				err = fmt.Errorf("bug commit: %v", err)
+				out <- core.NewImportError(err, "")
+				return
+			}
+		}
+		if searchIter.HasError() {
+			out <- core.NewImportError(searchIter.Err, "")
+		}
+	}()
+
+	return out, nil
+}
+
+// Create a bug.Person from a JIRA user
+func (self *jiraImporter) ensurePerson(
+	repo *cache.RepoCache, user User) (*cache.IdentityCache, error) {
+
+	// Look first in the cache
+	i, err := repo.ResolveIdentityImmutableMetadata(
+		keyJiraUser, string(user.Key))
+	if err == nil {
+		return i, nil
+	}
+	if _, ok := err.(entity.ErrMultipleMatch); ok {
+		return nil, err
+	}
+
+	i, err = repo.NewIdentityRaw(
+		user.DisplayName,
+		user.EmailAddress,
+		user.Key,
+		"",
+		map[string]string{
+			keyJiraUser: string(user.Key),
+		},
+	)
+
+	if err != nil {
+		return nil, err
+	}
+
+	self.out <- core.NewImportIdentity(i.Id())
+	return i, nil
+}
+
+// Create a bug.Bug based from a JIRA issue
+func (self *jiraImporter) ensureIssue(
+	repo *cache.RepoCache, issue Issue) (*cache.BugCache, error) {
+	author, err := self.ensurePerson(repo, issue.Fields.Creator)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO(josh)[f8808eb]: Consider looking up the git-bug entry directly from
+	// the jira field which contains it, if we have a custom field configured
+	// to store git-bug IDs.
+	b, err := repo.ResolveBugCreateMetadata(keyJiraID, issue.ID)
+	if err != nil && err != bug.ErrBugNotExist {
+		return nil, err
+	}
+
+	if err == bug.ErrBugNotExist {
+		cleanText, err := text.Cleanup(string(issue.Fields.Description))
+		if err != nil {
+			return nil, err
+		}
+
+		title := fmt.Sprintf("[%s]: %s", issue.Key, issue.Fields.Summary)
+		b, _, err = repo.NewBugRaw(
+			author,
+			issue.Fields.Created.Unix(),
+			title,
+			cleanText,
+			nil,
+			map[string]string{
+				keyOrigin:      target,
+				keyJiraID:      issue.ID,
+				keyJiraKey:     issue.Key,
+				keyJiraProject: self.conf[keyProject],
+			})
+		if err != nil {
+			return nil, err
+		}
+
+		self.out <- core.NewImportBug(b.Id())
+	} else {
+		self.out <- core.NewImportNothing("", "bug already imported")
+	}
+
+	return b, nil
+}
+
+// Return a unique string derived from a unique jira id and a timestamp
+func getTimeDerivedID(jiraID string, timestamp MyTime) string {
+	return fmt.Sprintf("%s-%d", jiraID, timestamp.Unix())
+}
+
+// Create a bug.Comment from a JIRA comment
+func (self *jiraImporter) ensureComment(
+	repo *cache.RepoCache, b *cache.BugCache, item Comment) error {
+	// ensure person
+	author, err := self.ensurePerson(repo, item.Author)
+	if err != nil {
+		return err
+	}
+
+	targetOpID, err := b.ResolveOperationWithMetadata(
+		keyJiraID, item.ID)
+	if err == nil {
+		self.out <- core.NewImportNothing("", "comment already imported")
```

Comment:
Those Nothing event are not really useful the way they are done at the moment ... (for all the bridges, not just this PR).

I think that ideally, the bridge would emit an event for each kind of imported thing (bug, comment, label ...) **or** the nothing event (meaning nothing happened on the bug) **or** an error event.

Reply:
I updated https://github.com/MichaelMure/git-bug/issues/231 about this.

Reply:
The goal is to have a stream of event that UIs can interpret, filter and display as they want. The "nothing" event say "we have this bug but nothing has changed, it's up to date". Emitting an event for "no title has been changed" would not be very useful.

Regarding your question, I'd say it depends on how sever the problem is. Does the user reasonably expect the import to work ? In that case that would be an error. If not ... discussing with @A-Hilaly we considered adding a warning event. Maybe that would be a good way to do that. The error would end the import process of a bug and optionally of the whole import, a warning would not.